fsspec/asyn.py

```diff
@@ -568,8 +568,8 @@ async def _put(
         coros = []
         callback.set_size(len(file_pairs))
         for lfile, rfile in file_pairs:
-            callback.branch(lfile, rfile, kwargs)
-            coros.append(self._put_file(lfile, rfile, **kwargs))
+            put_file = callback.branch_coro(self._put_file)
```

Comment:
I don't like adding `branch_coro` but I am not sure how to support scoped/structured callbacks in async without one either, since we are working task/futures.

Reply:
Yeah, this is not well explorer territory. I would have thought that all calls into the callbacks machinery can by sync, since they will be fast and so safe to call from sync or async contexts.